northstar/src/runtime/state.rs

```diff
@@ -109,615 +89,464 @@ impl ProcessContext {
+                return Err(Error::UnknownApplication(key.clone()));
+            }
+
+            // Check if the container is already started
+            if container.process.is_some() {
+                warn!("Application {} is already running", key);
+                return Err(Error::ApplicationStarted(key.clone()));
+            }
 
-        // Check for all required resources
-        for mount in app.container.manifest.mounts.values() {
-            if let Mount::Resource { name, .. } = mount {
-                if !resources.contains(name) {
-                    return Err(Error::MissingResource(name.clone()));
+            container.manifest.clone()
+        } else if self.npks.contains_key(key) {
+            let container = self.mount(&key).await?;
+            let manifest = container.manifest.clone();
+            mounted.push(key.clone());
+            self.containers.insert(key.clone(), container);
+            manifest
+        } else {
+            warn!("Unknown application {}", key);
+            return Err(Error::UnknownApplication(key.clone()));
+        };
+
+        // Find to be mounted resources
+        for key in manifest.mounts.values().filter_map(|m| match m {
+            Mount::Resource {
+                name,
+                version,
+                repository,
+                ..
+            } => Some(ContainerKey::new(
+                name.clone(),
+                version.clone(),
+                repository.clone(),
+            )),
+            _ => None,
+        }) {
+            // Mount not yet mounted resources
+            if !self.containers.contains_key(&key) {
+                if self.npks.contains_key(&key) {
+                    // Obtain the key from the repo - needed for mounting
+                    match self.mount(&key).await {
+                        Ok(container) => {
+                            // Fine. Add the container to the list of mounted containers
+                            mounted.push(key.clone());
+                            self.containers.insert(key, container);
+                        }
+                        Err(e) => {
+                            warn!("Failed to mount {}", key);
+                            // Umount everything mounted for this particular start operation
+                            for k in mounted.drain(..) {
+                                self.umount(&k).await.ok();
+                            }
+                            return Err(e);
+                        }
+                    }
+                } else {
+                    for k in mounted.drain(..) {
+                        self.umount(&k).await.ok();
+                    }
+                    return Err(Error::UnknownResource(key));
                 }
             }
         }
 
-        // Spawn process
-        info!("Starting {}", app);
+        // This must exist
+        let container = self.containers.get(key).unwrap();
```

Comment:
and no ugly "let's unwrap since it was already checked" code

Reply:
This is the result of a borrow checker fight. I couldn't solve it. It because above a `Container` needs to be inserted into `self.containers` which mutably borrows.

Can you find a solution?